### PR TITLE
Small speed-up in movepick

### PR DESCRIPTION
**Pull request by protonspring**
It seems to be a little faster if we remove the auto/begin/end code and explicitly go through the moves.

Master
cycles: 174,016,883,088
instructions: 262,385,659,925
time: 49.89 seconds

Patch
cycles: 173,061,453,047
instructions: 262,366,840,606
time: 49.56 seconds

### CODE REVIEW

**Comment by MichaelB7** [comment]
1% gain here.

**Comment by nguyenpham** [comment]
It is not clear to me. Did you test on other systems with other compilers? AFAIK, the original code is easier for compilers to optimize

**Comment by protonspring** [comment]
@nguyenpham Could you test it for us on your compiler and report some actual numbers?

**Comment by d3vv** [comment]
@protonspring Could you suggest - how to test for get such "cycles" and  "instructions" values?

@nguyenpham And AFAIK proton-patch is more simple for any level compiler optimizations..

**Comment by nguyenpham** [comment]
I can’t help atm since I have been struggling to compile sf with optimised mode (I am on an iMac and failed to compile with profile parameter).

BTW, from my experience, test out of 1% different speed is very hard since since it’s so close to a random

**Comment by d3vv** [comment]
@nguyenpham There is no problem to use PGO for mac, please see link below

https://github.com/official-stockfish/Stockfish/issues/2053

**Comment by vondele** [comment]
@d3vv I assume the cycle counter comes from the `sudo perf stat -r 5 -a -B -e cycles:u,instructions:u ./stockfish bench > /dev/null ` measurement as described here:
https://github.com/glinscott/fishtest/wiki/Creating-my-first-test#speed-optimization

**Comment by MichaelB7** [comment]
> I can’t help atm since I have been struggling to compile sf with optimised mode (I am on an iMac and failed to compile with profile parameter).
> 
> BTW, from my experience, test out of 1% different speed is very hard since since it’s so close to a random

Install clang or gcc from Macports .  Much easier.  Google Macports  all you need to do is there.  it's doable.  Also - Apple computers are not a great barometer on what works well on Windows or Linux.  Over 95% of Stockfish users are Windows or Linux ( based on 20,000 downloads of my Stockfish exe's).

File: src/movepick.cpp [hunk 1]
```diff
@@ -99,13 +99,15 @@ void MovePicker::score() {
 
   static_assert(Type == CAPTURES || Type == QUIETS || Type == EVASIONS, "Wrong type");
 
-  for (auto& m : *this)
+  for (ExtMove *p = cur; p < endMoves; ++p)
```
Comment (skiminki) [inline src/movepick.cpp:102]: I'm curious about what is the cause of the speedup here. Is it the changed for loop?

**Comment by protonspring** [comment]
The loop is exactly the same. Just less code to do it.

**Comment by skiminki** [comment]
> The loop is exactly the same. Just less code to do it.

Yes, but I was curious on whether you checked the disassembly. TBH, this looks like a missed optimization in GCC and/or CLANG, if this gives any performance boost. Internally, the compiler should convert "for (auto &m: *this)" to more or less to the explicit form. Probably, the actual compiler-generated form is "for (ExtMove *p = cur; p != endMoves; ++p)" but it shouldn't matter to the CPU whether we use '<' or '!='

**Comment by protonspring** [comment]
I simply noticed that it is faster on my machines.  You should check the
disassembly.  It would be interesting to see.

**Comment by Sopel97** [comment]
no difference in assembly for all GenTypes in gcc 10.1
https://godbolt.org/z/7-CJ9g

**Comment by skiminki** [comment]
@protonspring What compiler are you using?

**Comment by protonspring** [comment]
I think it's gcc 7.5.0

**Comment by skiminki** [comment]
I have the opposite effect on my machine (gcc 9.2.0):

**Two runs with SF vanilla:**

~/projects-git/Stockfish/src % sudo perf stat -r 5 -a -B -e cycles:u,instructions:u ./sf.sh                                                  

 Performance counter stats for 'system wide' (5 runs):

    24,399,340,806      cycles:u                                                      ( +-  0.06% )
    40,569,610,081      instructions:u            #    1.66  insn per cycle           ( +-  0.02% )

           5.21708 +- 0.00604 seconds time elapsed  ( +-  0.12% )

~/projects-git/Stockfish/src % sudo perf stat -r 5 -a -B -e cycles:u,instructions:u ./sf.sh

 Performance counter stats for 'system wide' (5 runs):

    24,429,509,455      cycles:u                                                      ( +-  0.08% )
    40,574,181,974      instructions:u            #    1.66  insn per cycle           ( +-  0.01% )

           5.22019 +- 0.00343 seconds time elapsed  ( +-  0.07% )



**Two runs with your patch:**

~/projects-git/Stockfish/src % sudo perf stat -r 5 -a -B -e cycles:u,instructions:u ./sf.sh                                                  

 Performance counter stats for 'system wide' (5 runs):

    24,567,163,109      cycles:u                                                      ( +-  0.06% )
    40,603,453,292      instructions:u            #    1.65  insn per cycle           ( +-  0.00% )

           5.25169 +- 0.00502 seconds time elapsed  ( +-  0.10% )

~/projects-git/Stockfish/src % sudo perf stat -r 5 -a -B -e cycles:u,instructions:u ./sf.sh

 Performance counter stats for 'system wide' (5 runs):

    24,572,186,976      cycles:u                                                      ( +-  0.04% )
    40,603,018,129      instructions:u            #    1.65  insn per cycle           ( +-  0.00% )

           5.26209 +- 0.00490 seconds time elapsed  ( +-  0.09% )

(sf.sh = "exec ./stockfish bench 16 1 16 >/dev/null 2>&1" so that there's absolutely nothing on terminal for cleanest possible runs)

**Comment by skiminki** [comment]
A similar trend is with `./stockfish bench | tail -n 3`. There's about 0.5% nps difference in average (6 runs, vanilla min > your patch max). This is in favor of vanilla over your patch.

Either GCC 9 has fixed sometime since 7.5, or there's something other subtle happening like different program block ordering. I'll quickly test with CLANG, too

**Comment by skiminki** [comment]
CLANG 9.0.1 favors your patch a bit: 
~/projects-git/Stockfish/src % sudo perf stat -r 5 -a -B -e cycles:u,instructions:u ./sf.sh 

Vanilla
```
 Performance counter stats for 'system wide' (5 runs):

     7,513,979,811      cycles:u                                                      ( +-  0.05% )
    12,783,453,828      instructions:u            #    1.70  insn per cycle           ( +-  0.05% )

           1.62106 +- 0.00132 seconds time elapsed  ( +-  0.08% )

~/projects-git/Stockfish/src % sudo perf stat -r 5 -a -B -e cycles:u,instructions:u ./sf.sh                                                  

 Performance counter stats for 'system wide' (5 runs):

     7,512,815,894      cycles:u                                                      ( +-  0.08% )
    12,794,024,870      instructions:u            #    1.70  insn per cycle           ( +-  0.13% )

           1.61740 +- 0.00210 seconds time elapsed  ( +-  0.13% )
```

Your patch:

```
~/projects-git/Stockfish/src % sudo perf stat -r 5 -a -B -e cycles:u,instructions:u ./sf.sh                                                  

 Performance counter stats for 'system wide' (5 runs):

     7,502,794,425      cycles:u                                                      ( +-  0.04% )
    12,790,076,365      instructions:u            #    1.70  insn per cycle           ( +-  0.01% )

          1.613580 +- 0.000672 seconds time elapsed  ( +-  0.04% )

~/projects-git/Stockfish/src % sudo perf stat -r 5 -a -B -e cycles:u,instructions:u ./sf.sh                                                  

 Performance counter stats for 'system wide' (5 runs):

     7,503,831,715      cycles:u                                                      ( +-  0.08% )
    12,790,691,450      instructions:u            #    1.70  insn per cycle           ( +-  0.00% )

           1.61615 +- 0.00281 seconds time elapsed  ( +-  0.17% )
```

**Comment by MichaelB7** [comment]
Just a general comment - bench runs , my view should be done with PGO builds and it’s probably best to them to least depth 20.  It’s hard to imagine that running a test on 50 or so positions that less than 2 seconds in total is very worth while.  Others may have a difference in opinion which I would be interested in.

**Comment by skiminki** [comment]
I'm running my runs with "profile-build", so PGO is enabled. But I agree, larger depths should be used that what I used.

**Comment by skiminki** [comment]
Theo posted disassembly diffs for gcc-8 between those 2 patches in TCEC discord. They don't really support the case for this patch being a win or loss, unless there's something specific in how CPUs handle the different jumps. It is possible that there is a difference in static branch prediction.

https://cdn.discordapp.com/attachments/660216836922540052/721029054303830116/movepick-gcc8.diff

**Comment by skiminki** [comment]
`taskset -c 4 stockfish bench 32 1 21` results: (GCC-9.2.0, core-i7 8700k, fixed CPU core)

**Vanilla**
```
 Performance counter stats for 'system wide' (5 runs):

   126,462,532,617      cycles:u                                                      ( +-  0.07% )
   206,827,614,712      instructions:u            #    1.64  insn per cycle           ( +-  0.00% )

           26.9330 +- 0.0180 seconds time elapsed  ( +-  0.07% )
```

**This PR**
```
 Performance counter stats for 'system wide' (5 runs):

   127,062,574,427      cycles:u                                                      ( +-  0.04% )
   206,958,738,885      instructions:u            #    1.63  insn per cycle           ( +-  0.00% )

          27.06565 +- 0.00832 seconds time elapsed  ( +-  0.03% )
```

On my box, that's 0.47% more cycles and 0.49% more time.

**Comment by protonspring** [comment]
@skiminki Just curious how you are so confused by this PR.  It tests faster on my machines, and I posted it for others to verify.  Instead of 50 posts, it would be more helpful to simply post your benchmarks.

**Comment by MichaelB7** [comment]
@skiminko - I thought you had a Threadripper - I’m got a 1% Pickup  with the 3970x - hard to believe that is mostly noise.  gcc 10.1 v

**Comment by skiminki** [comment]
Ok, let me summarize my 50 previous posts. Couple of the first ones are about trying to understand whether the patch actually does something or not. One of the patches is about posting a link to theo's assembly diff which confirms that the only effect in the binary is the different comparison i.e. different conditional jumps.

4 of the posts discuss the benchmarks directly: 2x GCC 9 (small and med depth) both disfavoring your patch. 1x CLANG bench which favors your patch. These are all with numbers and all with PGO. (The fourth isn't with numbers, because I didn't feel like it's worth the effort to calculate avg + error bars manually from reported nps.)

I'm confused because I have hard time understanding what causes the speed difference. Also, my bench results don't really corroborate your findings, but they don't refute them either.

Rather than noise, I would think there's something systematic in the GCC build that causes these differences, possibly due to cache effects or similar. I don't know. But it feels that this patch doesn't directly improve the perf, but rather, makes a lucky stirring in the way the program gets linked. Another similar patch might have the opposite effect.

There is a possibility that this patch actually changes something, but then that would have to be about how the CPUs do static branch prediction for different jumps. (Static meaning that there is not yet dynamic history for the specific branch in the table.)

Note that I'm not saying that this patch shouldn't get merged, but I'm suspicious of the performance claims.

> I thought you have a Threadripper - I’m got a 1% Pickup with the 3970x -
> hard to believe that is mostly noise. gcc 10.1 v

I wish I had a Threadripper :)

**Comment by MichaelB7** [comment]
@skiminki I do not disagree with any of your points.  Your concerns are vaild.

**Comment by protonspring** [comment]
I don't know what a ThreadRipper is, but it sure sounds manly.

**Comment by skiminki** [comment]
I may have an explanation. Reference: https://software.intel.com/content/www/us/en/develop/download/intel-64-and-ia-32-architectures-optimization-reference-manual.html

Section 3.4.2.2 Optimizing for Macrofusion

This section lists various combinations of instructions pairs (e.g., cmp + jne) that can be fused in Haswell+. Fusion here means that two instructions will get merged into a single µop that combines both these assembler instructions (= faster perf). On Haswell+, jne and je support all fusion options, whereas the other conditional jumps support only some. This could actually explain why there's a perf difference even when only the jumps are changed.

It is entirely plausible that AMD implements a different set of macrofusion combinations. This could explain why your patch improves performance on AMD boxes but decreases performance on my Intel box.

**Comment by sqrmax** [comment]
> Theo posted disassembly diffs for gcc-8 between vanilla/this patch in TCEC discord. They don't really support the case for this patch being a win or loss, unless there's something specific in how CPUs handle the different jumps. It is possible that there is a difference in static branch prediction.
> 
> https://cdn.discordapp.com/attachments/660216836922540052/721029054303830116/movepick-gcc8.diff

Based on this diff, I do not see grounds for a claim of less code.  I would be really interested in seeing an assembly diff for the compiler in which this patch was developed.

**Comment by sqrmax** [comment]
> I simply noticed that it is faster on my machines. You should check the disassembly. It would be interesting to see.

Do you have the disassembly for your compiler handy?  I do not have access to your system to see what happened.  What CPU is involved here?  What does gcc --version output for you?

**Comment by sqrmax** [comment]
Looking at skiminki's diff,

https://cdn.discordapp.com/attachments/660216836922540052/721029054303830116/movepick-gcc8.diff

consider this instruction selection pair:

@@ -700,7 +700,7 @@
     movq    72(%rbx), %rsi
     movq    %rax, 80(%rbx)
     cmpq    %rsi, %rax
-    je    .L77
+    jbe    .L77

Those two jumps are not mere opposites of each other, they mean different things.  That happens elsewhere as well, e.g.:

@@ -662,8 +662,8 @@
 	movq	%rax, %rsi
 .L73:
 	movq	72(%rbx), %rcx
-	cmpq	%rsi, %rcx
-	jnb	.L183
+	cmpq	%rcx, %rsi
+	jbe	.L183

Let's assume all the code is semantically correct.  At least in theory, the different jumps allow for different dynamic behavior of how control goes between basic blocks.  If that's the case, the improvement might not visible in terms of an assembly diff, one has to look at what the code is doing.  The easy case would be that, because of the types of variables involved in the comparison, the jumps are functionally equivalent.  In that case, I'd like to believe PGO should take care of these things automatically.

**Comment by skiminki** [comment]
Ok, I spent some time looking at the disassembler diffs (they're actually much more for PGO builds than Theo's original diff). Pondering about what could explain those diffs, we came to the conclusion with sqrmax that perhaps '<' as a loop predicate instead of '!=' can actually improve the code. The reason is that the compiler can use the Duff device with inlined loops when using '<', but this is not true when using '!='.

I also measured branch mispredictions and there was a 0.9% reduction with this PR which would support more efficient loop inlining with less branching.

It is also a possibility that my box suffers from some kind of secondary issue with this particular change resulting in a slightly different memory layout. This could explain why I get better code with CLANG (no secondary issues by chance) and worse code with GCC (triggers something such as L1 set conflict or whatever by chance.)

So maybe there's a slight improvement potential with this patch, but for quite unobvious reasons.

The above is theoretical at this point and I'm not planning to spend the whole weekend proving/disproving this. :)

**Comment by protonspring** [comment]
@skiminki Thank you for spending so much time looking at this.  I don't have that much patience, but I appreciate your analysis.

**Comment by snicolet** [comment]
I tested this patch for speed on my system, using two long benches ran on parallel
(the script I use is there: http://cassio.free.fr/stockfish/speed )

The patch didn't bring any speed-up compared to master:

```
===========================
master
Total time (ms) : 55834
Nodes searched  : 105059398
Nodes/second    : 1881638

===========================
testedpatch
Total time (ms) : 55963
Nodes searched  : 105059398
Nodes/second    : 1877301

```

Hope this helps, Stéphane

**Comment by protonspring** [comment]
since there are mixed results. i will close.

**Pull request closed without merging**